src/workerd/jsg/modules.c++

```diff
@@ -322,32 +334,43 @@ NonModuleScript NonModuleScript::compile(kj::StringPtr code, jsg::Lock& js, kj::
   return NonModuleScript(js, check(v8::ScriptCompiler::CompileUnboundScript(isolate, &source)));
 }
 
-void instantiateModule(jsg::Lock& js, v8::Local<v8::Module>& module) {
+void instantiateModule(
+    jsg::Lock& js, v8::Local<v8::Module>& module, InstantiateModuleOptions options) {
   KJ_ASSERT(!module.IsEmpty());
   auto isolate = js.v8Isolate;
   auto context = js.v8Context();
 
   auto status = module->GetStatus();
-  // Nothing to do if the module is already instantiated, evaluated, or errored.
-  if (status == v8::Module::Status::kInstantiated || status == v8::Module::Status::kEvaluated ||
-      status == v8::Module::Status::kErrored)
-    return;
 
-  JSG_REQUIRE(status == v8::Module::Status::kUninstantiated, Error,
-      "Module cannot be synchronously required while it is being instantiated or evaluated. "
-      "This error typically means that a CommonJS or NodeJS-Compat type module has a circular "
-      "dependency on itself, and that a synchronous require() is being called while the module "
-      "is being loaded.");
+  // If the previous instantiation failed, throw the exception.
+  if (status == v8::Module::Status::kErrored) {
+    isolate->ThrowException(module->GetException());
+    throw jsg::JsExceptionThrown();
+  }
+
+  // Nothing to do if the module is already evaluated.
+  if (status == v8::Module::Status::kEvaluated || status == v8::Module::Status::kEvaluating) return;
+
+  if (status == v8::Module::Status::kUninstantiated) {
+    jsg::check(module->InstantiateModule(context, &resolveCallback));
+  }
 
-  jsg::check(module->InstantiateModule(context, &resolveCallback));
   auto prom = jsg::check(module->Evaluate(context)).As<v8::Promise>();
+
+  if (module->IsGraphAsync() && prom->State() == v8::Promise::kPending) {
+    // If top level await has been disable, error.
+    JSG_REQUIRE(options != InstantiateModuleOptions::NO_TOP_LEVEL_AWAIT, Error,
+        "Top-level await in module is not permitted at this time.");
+  }
```

Comment:
This can be simplified to a single line of JSG_REQUIRE but I guess it would be less readable...

Reply:
Would it be more readable and maintainable if we add these assumptions/assertions as a comment before the function? It seems there are lots of know-how required for this function that could be beneficial for a developer, if provided in a more structural way.